app/android/runtime_client/src/org/xwalk/app/runtime/extension/JsStubGenerator.java

```diff
@@ -0,0 +1,116 @@
+package org.xwalk.app.runtime.extension;
+
+import android.util.Log;
+import java.util.Map;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import org.xwalk.app.runtime.extension.ReflectionHelper.MemberInfo;
+
+public class JsStubGenerator {
+    static public String TAG = "JsStubGenerator";
+	ReflectionHelper reflection;
+	String jsHeader =
+        "var jsStub = requireNative(\"jsStub\").jsStub;\n" +
+        "var helper = jsStub.create(exports, extension);\n";
+	String jsEnder = "\n";
+	
+	JsStubGenerator (ReflectionHelper mirror) {
+		reflection = mirror;
+	}
+
+	String generate() {
+		String result = "";
+        if(reflection.getEventList() != null) {
+            result += generateEventTarget();
+        }
+
+		Map<String, MemberInfo> members = reflection.getMembers();
+		for (String key : members.keySet()) {
+			MemberInfo m = members.get(key);
+            switch (m.type) {
+                case JS_PROPERTY:
+                    result += generateProperty(key, m);
+                    break;
+                case JS_METHOD:
+                    result += generateMethod(key, m);
+                    break;
+                default:
+                    break;
+            }
+		}
+		return jsHeader + result + jsEnder;
+	}
+
+	String generateEventTarget() {
+        String[] eventList = reflection.getEventList();
+        if(eventList == null || eventList.length < 1) {
+            return "";
+        }
+
+        String gen = "jsStub.makeEventTarget(exports);\n";
+        for(String e : eventList) {
+            gen += "helper.addEvent(\"" + e + "\");\n";
+        }
+        return gen;
+	}
+	
+	String generateProperty(String name, MemberInfo m) {
+        if(m.isWritable) {
+            return "jsStub.defineProperty(exports, \"" + name + "\", true);\n";
+        } else {
+            return "jsStub.defineProperty(exports, \"" + name + "\");\n";
+        }
+	}
+
+    String generatePromiseMethod(String name, String jsArgs) {
+        String argStr = "{\"resolve\": resolve, \"reject\":reject}";
+        if(jsArgs.length() > 0) {
+            argStr = jsArgs + ", " + argStr;
+        }
+        return String.format(
+                "exports.%s = function(%s) {\n" +
+                "  return new Promise(function(resolve, reject){\n" +
```

Comment:
No, it should be the native implement in V8. I think sysapps_promise is implemented because the native one is not ready that time.